### PR TITLE
Adding metrics to Swift Jobs

### PR DESCRIPTION
**Pull request by thoven87**
Here's my first attempt at adding Observability to the Swift jobs project.

I would love some feedback on how to make metrics better.  cc @adam-fowler 

Since everyone might use different tools to display their dashboards, I will attempt to provide a grafana dashboard as it is what I am most familiar with.

Lastly, most of the tests are a compilation of the current job tests and [Hummingbird](https://github.com/hummingbird-project/hummingbird/blob/main/Tests/HummingbirdTests/MetricsTests.swift) metrics tests.

### CODE REVIEW

File: Sources/Jobs/JobQueue.swift [hunk 1]
```diff
@@ -41,11 +42,17 @@ public struct JobQueue<Queue: JobQueueDriver>: Service {
...
     /// - Returns: Identifier of queued job
-    @discardableResult public func push<Parameters: Codable & Sendable>(id: JobIdentifier<Parameters>, parameters: Parameters) async throws -> Queue.JobID {
+    @discardableResult public func push<Parameters: Codable & Sendable>(
+        id: JobIdentifier<Parameters>, parameters: Parameters
+    ) async throws -> Queue.JobID {
         let jobRequest = JobRequest(id: id, parameters: parameters)
         let buffer = try JSONEncoder().encodeAsByteBuffer(jobRequest, allocator: self.allocator)
         let id = try await self.queue.push(buffer)
-        self.handler.logger.debug("Pushed Job", metadata: ["_job_id": .stringConvertible(id), "_job_type": .string(jobRequest.id.name)])
+        Counter(label: "queued_jobs_counter", dimensions: [("queue_name", jobRequest.id.name)]).increment()
```
Comment (adam-fowler) [inline Sources/Jobs/JobQueue.swift:51]: queue_name is a little misleading, shouldn't this be job_name, or JobName to coincide with the log

File: Sources/Jobs/JobQueueHandler.swift [hunk 2]
```diff
@@ -119,4 +126,40 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
...
+            label: "\(name)_job_duration",
+            dimensions: [
+                ("success", error == nil ? "true" : "false"),
+                ("id", name),
+            ],
+            preferredDisplayUnit: .seconds
+        ).recordNanoseconds(DispatchTime.now().uptimeNanoseconds - startTime)
+
+        // Increment job counter base on status
+        if error != nil {
```
Comment (adam-fowler) [inline Sources/Jobs/JobQueueHandler.swift:146]: if let error

File: Sources/Jobs/JobQueueHandler.swift [hunk 3]
```diff
@@ -119,4 +126,40 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
...
+            if error is CancellationError {
+                Counter(
+                    label: "cancelled_jobs_counter",
+                    dimensions: [("queue_name", name)]
```
Comment (adam-fowler) [inline Sources/Jobs/JobQueueHandler.swift:150]: Again queue_name, should this not be JobName

File: Sources/Jobs/JobQueueHandler.swift [hunk 4]
```diff
@@ -119,4 +126,40 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
...
     public var description: String { "JobQueueHandler<\(String(describing: Queue.self))>" }
+
+    private func updateJobMetrics(
+        for name: String,
+        startTime: UInt64,
+        error: Error? = nil
+    ) {
+        // Calculate job execution time
+        Timer(
+            label: "\(name)_job_duration",
```
Comment (adam-fowler) [inline Sources/Jobs/JobQueueHandler.swift:137]: Shouldn't the name just be the a dimension. It is elsewhere

File: Tests/JobsTests/MetricsTests.swift [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+import Atomics
+import Jobs
+import Logging
+import Metrics
+import NIOConcurrencyHelpers
+import ServiceLifecycle
+import XCTest
+
+final class TestMetrics: MetricsFactory {
```
Comment (adam-fowler) [inline Tests/JobsTests/MetricsTests.swift:23]: I've just made some changes to this to make it compile in Swift 6 mode. You'll need to wrap all this mutable state in NIOLockedValueBoxes. If you just copied and pasted this from Hummingbird, you can probably do the same again. Although you will need to edit the accesses in the `XCTAssert...` calls

File: Package.swift [hunk 6]
```diff
@@ -28,6 +29,7 @@ let package = Package(
                 .product(name: "NIOCore", package: "swift-nio"),
                 .product(name: "NIOFoundationCompat", package: "swift-nio"),
                 .product(name: "ServiceLifecycle", package: "swift-service-lifecycle"),
+                .product(name: "Metrics", package: "swift-metrics"),
```
Comment (adam-fowler) [inline Package.swift:32]: Minor nit, I added these in alphabetical order

**Review by adam-fowler — COMMENTED** [review]
Here are some initial comments. In general this look good, I was wondering if it is worthwhile adding a metric for the retry count.

For another PR if we added a dateCreated to the QueuedJob, we could add a time job queued metric. In a way dateCreated is probably a useful piece data regardless.

File: Tests/JobsTests/MetricsTests.swift (hunk 5 above)
Comment (thoven87) [reply Tests/JobsTests/MetricsTests.swift:23]: > I've just made some changes to this to make it compile in Swift 6 mode. You'll need to wrap all this mutable state in NIOLockedValueBoxes. If you just copied and pasted this from Hummingbird, you can probably do the same again. Although you will need to edit the accesses in the `XCTAssert...` calls

Thanks Adam! I have updated the helpers with the help of your previous PR.

**Comment by thoven87** [comment]
> Here are some initial comments. In general this look good, I was wondering if it is worthwhile adding a metric for the retry count.
> 
> For another PR if we added a dateCreated to the QueuedJob, we could add a time job queued metric. In a way dateCreated is probably a useful piece data regardless.

Thanks, for the feedback. I added a retry counter. I will open another PR to add both `createdAt` and maybe Job name? Right now, `Job.id.name` does not always exist in every context. In some context, it's just Job.id which is a UUID.

File: Sources/Jobs/JobQueueHandler.swift [hunk 7]
```diff
@@ -64,7 +67,8 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
 
     func runJob(_ queuedJob: QueuedJob<Queue.JobID>) async throws {
         var logger = logger
-        logger[metadataKey: "_job_id"] = .stringConvertible(queuedJob.id)
+        let startTime = DispatchTime.now().uptimeNanoseconds
+        logger[metadataKey: "JobId"] = .stringConvertible(queuedJob.id)
```
Comment (Joannis) [inline Sources/Jobs/JobQueueHandler.swift:71]: Here you're making each word Uppercased, whereas the other key is snake_cased. I'd prefer sticking to one.

**Comment by Joannis** [comment]
Looks great! Just one bit of consistency

File: Sources/Jobs/JobQueueHandler.swift (hunk 7 above)
Comment (thoven87) [reply Sources/Jobs/JobQueueHandler.swift]: > Here you're making each word Uppercased, whereas the other key is snake_cased. I'd prefer sticking to one.

Agreed! Fixed in the latest commit.

**Comment by thoven87** [comment]
I think this Pull request is ready. Below is a dashboard I created to test this implementation.

Next step would be to add tracing and queued at to compute avg wait time.
Screenshot 2024-08-18 at 12 35 45 PM

Screenshot 2024-08-18 at 12 50 28 PM

**Comment by thoven87** [comment]
also previous implementation of the dashboard to show the failed metrics
Screenshot 2024-08-17 at 8 46 21 PM

File: Sources/Jobs/JobQueueHandler.swift [hunk 8]
```diff
@@ -119,4 +130,70 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
...
+        // Increment job counter base on status
+        Counter(
+            label: self.metricsLabel,
+            dimensions: dimensions
+        ).increment()
+        
+        self.updateJobMeters()
+    }
+    
+    private func updateJobMeters() {
```
Comment (adam-fowler) [inline Sources/Jobs/JobQueueHandler.swift:194]: Can we rename this. It isn't clear what it is doing. Also if this is called if the job fails to decode, or it couldn't find the job id then processing is decremented when it hasn't been incremented.

I'm wondering if it be easier to break out the queued and processing counter editing from here and just place at the top of `runJob`. You'd also have to remove the processing `decrement` when you retry a job. 

```swift
Meter(label: self.meterLabel, dimensions: [("status", JobStatus.queued.rawValue)]).decrement()
Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).increment()
defer {
    Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).decrement()
}

Comment (thoven87) [reply Sources/Jobs/JobQueueHandler.swift:194]: > Can we rename this. It isn't clear what it is doing. Also if this is called when the job fails to decode, or it couldn't find the job id then processing is decremented when it hasn't been incremented.
> 
> I'm wondering if it be easier to break out the queued and processing counter editing from here and just place at the top of `runJob`. You'd also have to remove the processing `decrement` when you retry a job.
> 
> ```swift
> Meter(label: self.meterLabel, dimensions: [("status", JobStatus.queued.rawValue)]).decrement()
> Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).increment()
> defer {
>     Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).decrement()
> }
> ```

This is much cleaner! I will update accordingly, thanks!

Comment (thoven87) [reply Sources/Jobs/JobQueueHandler.swift]: @adam-fowler I updated the PR with the suggestion above.

**Comment by thoven87** [comment]
> This is great

Please merge at your convenience! What's the best place to share RED metrics dashboards? I am in the process of creating a few  Grafana dashboards for Hummingbird.

There would be three dashboards in total: 

1 -  Display the overall health of an application including jobs and http metrics, spans and error logs. 
2 - Display Http metrics, spans and error logs
3 - Display Jobs metrics, spans and error logs

**Comment by adam-fowler** [comment]
Just waiting on a final approval from @Joannis

**Review by Joannis — APPROVED** [review]
Looks awesome!

**Comment by Joannis** [comment]
@thoven87 I'd love to post them either in a docs section, on docs.hummingbird.codes as a DocC guide. I think we can also highlight it on swiftonserver.com

**Comment by thoven87** [comment]
> @thoven87 I'd love to post them either in a docs section, on docs.hummingbird.codes as a DocC guide. I think we can also highlight it on swiftonserver.com

@Joannis I suppose for DooC guide it would need to be added in a section in [this repository](https://github.com/hummingbird-project/hummingbird-docs) Indeed, it would be great to highlight it on swiftonserver.com as well.

**Comment by thoven87** [comment]
> You could possibly include something in the observability docs but https://docs.hummingbird.codes/2.0/documentation/hummingbird/loggingmetricsandtracing.
> 
> I think an end to end metrics and tracing article in swiftonserver.com which uses these would be a great addition.

Thanks Adam, I will take a look and see what I can do.

**Pull request merged**